src/Http/Routing/src/RouteOptions.cs

```diff
@@ -152,8 +156,20 @@ public void SetParameterPolicy(string token, [DynamicallyAccessedMembers(Dynamic
         _constraintTypeMap[token] = type;
     }
 
-    private static void AddConstraint<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TConstraint>(Dictionary<string, Type> constraintMap, string text) where TConstraint : IRouteConstraint
+    private static void AddConstraint<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TConstraint>(IDictionary<string, Type> constraintMap, string text) where TConstraint : IRouteConstraint
     {
         constraintMap[text] = typeof(TConstraint);
     }
 }
+
+internal class RegexErrorStubRouteConstraint : IRouteConstraint
+{
+    internal RegexErrorStubRouteConstraint([StringSyntax(StringSyntaxAttribute.Regex, RegexOptions.CultureInvariant | RegexOptions.IgnoreCase)] string regexPattern)
+    {
+    }
+
+    bool IRouteConstraint.Match(HttpContext? httpContext, IRouter? route, string routeKey, RouteValueDictionary values, RouteDirection routeDirection)
+    {
+        throw new InvalidOperationException(Resources.FormatRegexRouteContraint_NotConfigured(routeKey));
```

Comment:
Throwing the error should be in the constructor. That way, someone gets the error with any first request to their app.

If throwing the error is in the match then the error is delayed to the first request to that route.

Reply:
I've changed the implementation, but I'm not actually sure it is better. The root cause of the is in a few nested exceptions. Here is the example assertion:

```csharp
        Assert.IsType<RouteCreationException>(ex);
        Assert.IsType<InvalidOperationException>(ex.InnerException.InnerException);
        Assert.Equal(
            "A route parameter uses the regex constraint, which isn't registered. If this application was configured using CreateSlimBuilder(...) or AddRoutingCore(...) then this constraint is not registered by default. To use the regex constraint, configure route options at app startup: services.Configure<RouteOptions>(options => options.SetParameterPolicy<RegexInlineRouteConstraint>(\"regex\"));",
            ex.InnerException.InnerException.Message);
```

.... there are definitely benefits to throwing early to help folks avoid getting a nasty surprise in production if their test coverage isn't great.

Reply:
I think it's fine. Error logging includes inner exceptions, and people know to keep looking down for the real cause of the error. And the problem is related to creating the route anyway so the top error isn't inaccurate.